src/osd/PeeringState.cc

```diff
@@ -5575,13 +5575,30 @@ boost::statechart::result PeeringState::Active::react(const AdvMap& advmap)
     ps->share_pg_info();
   }
 
+  bool need_acting_change = false;
   for (size_t i = 0; i < ps->want_acting.size(); i++) {
     int osd = ps->want_acting[i];
     if (!advmap.osdmap->is_up(osd)) {
       pg_shard_t osd_with_shard(osd, shard_id_t(i));
-      ceph_assert(ps->is_acting(osd_with_shard) || ps->is_up(osd_with_shard));
+      if (!ps->is_acting(osd_with_shard) && !ps->is_up(osd_with_shard)) {
+        need_acting_change = true;
+      }
     }
   }
+  if (need_acting_change) {
+     psdout(10) << "Active need acting change, call choose_acting again"
+                << dendl;
+    // possibly because we re-add some strays into the acting set and
+    // some of them then go down in a subsequent map before we could see
+    // the map changing the pg temp.
+    // call choose_acting again to clear them out.
+    // note that we leave restrict_to_up_acting to false in order to
+    // not overkill any chosen stray that is still alive.
+    pg_shard_t auth_log_shard;
+    bool history_les_bound = false;
+    ps->remove_down_peer_info(advmap.osdmap);
```

Comment:
This should already be happening elsewhere, I think.

Reply:
Yeah, looks like in Started::react(const AdvMap&) -- note that Active ::react(const AdvMap&) returns forward_event() at the end so the parent handler will also fire.

Reply:
hmm, we need to move that down stray peer out of peer_info first, otherwise choose_acting would probably return the same want_acting set.

Reply:
hmm, we need to move that down stray peer out of peer_info first, otherwise choose_acting would probably return the same want_acting set, right?

Reply:
Hmm.  I suppose it's harmless to do it twice.